File: schemas/helpdesk-customer-support/convex/ticket_activities.ts
// ticket_activities: append-only audit trail of ticket changes for accountability and SLA debugging.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const ticket_activities = defineTable({
  ticketId: v.id("tickets"),
  userId: v.optional(v.id("users")),
  action: v.union(
    v.literal("created"),
    v.literal("updated"),
    v.literal("status_changed"),
    v.literal("priority_changed"),
    v.literal("assigned"),
    v.literal("escalated"),
    v.literal("reopened"),
    v.literal("resolved"),
    v.literal("closed"),
    v.literal("sla_breached")
  ),
  field: v.optional(v.string()),
  oldValue: v.optional(v.string()),
  newValue: v.optional(v.string()),
})
  .index("by_ticket_id_and_creation_time", ["ticketId"])
  .index("by_user_id", ["userId"]);

File: schemas/helpdesk-customer-support/spacetimedb/ticket_activities.rs
// ticket_activities: Append-only audit trail of ticket changes for accountability and SLA debugging.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum TicketActivityAction {
    Created,
    Updated,
    StatusChanged,
    PriorityChanged,
    Assigned,
    Escalated,
    Reopened,
    Resolved,
    Closed,
    SlaBreached,
}

#[spacetimedb::table(name = ticket_activities, public)]
pub struct TicketActivity {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub ticket_id: String, // UUID — FK → tickets.id (cascade delete)
    #[index(btree)]
    pub user_id: Option<String>, // UUID — FK → users.id (set null)
    pub action: TicketActivityAction,
    pub field: Option<String>,
    pub old_value: Option<String>,
    pub new_value: Option<String>,
    pub created_at: Timestamp,
    // Composite index: (ticket_id, created_at) — enforce in reducer logic
}
